PR: Display plots from collections editor in the Plots pane

jitseniesen opened the pull request:
<!--- Make sure to read the Contributing Guidelines:                   --->
<!--- https://github.com/spyder-ide/spyder/blob/master/CONTRIBUTING.md --->
<!--- and follow PEP 8, PEP 257 and Spyder's code style:               --->
<!--- https://github.com/spyder-ide/spyder/wiki/Dev:-Coding-Style      --->

## Description of Changes

* [x] Wrote at least one-line docstrings (for any new functions)
* [x] Added unit test(s) covering the changes (if testable)
* [x] Included a screenshot or animation (if affecting the UI, see [Licecap](https://www.cockos.com/licecap/))

<!--- Explain what you've done and why --->

The collection editor is opened when the user edits a list, set, dict or tuple in the Variable Explorer. At the moment, plotting in the collection editor opens a new window with the plot. This PR changes this so that the plot is displayed in the Plots pane, for consistency with the namespace browser. The Plots pane is brought to the front so that users notice the new plot. However, if the user unticks the setting "Mute inline plots" in the Plots pane, then the old behaviour is restored and plots are shown in a separate window.

The collection editor is also used in several places outside the Variable Explorer:
* The IPython Console uses it to show environment variables and sys.path contents. Neither of these can be plotted as far as I can see.
* The `oedit()` function can create a collection editor. I am not sure about this function. I don't think it is used anywhere in Spyder but perhaps it is meant to be run from the Internal Console. Plotting from a collection editor created by `oedit()` will not work, because the signal is not connected to anything.

One niggle is that plots from the collection editor will be shown in the currently active figure browser in the Plots pane, which is associated to the currently active console. However, the variable shown in the collection editor may originate from a different console.

![collection-plot](https://github.com/spyder-ide/spyder/assets/7941918/0f3cd8a6-9071-48a0-bd5d-d0b33471cd1c)


### Issue(s) Resolved

<!--- List the issue(s) below, in the form "Fixes #1234"; one per line --->

Fixes #21145


### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

<!--- TYPE YOUR USER/NAME AFTER THE FOLLOWING: --->
I certify the above statement is true and correct: Jitse Niesen

<!--- Thanks for your help making Spyder better for everyone! --->

ccordoba12 commented on spyder/plugins/variableexplorer/widgets/objectexplorer/toggle_column_mixin.py:
```diff
@@ -155,12 +155,21 @@ class ToggleColumnTreeView(QTreeView, ToggleColumnMixIn):
     show/hide columns.
     """
 
+    """
+    Emitted to request that a figure be shown.
+    First argument is the image, second argument is the mime type.
+    """
+    sig_show_figure_requested = Signal(bytes, str)
```
```suggestion
    sig_show_figure_requested = Signal(bytes, str)
    """
    This is emitted to request that a figure be shown.
    
    Parameters
    ----------
    image: bytes
        The image to show.
    mime_type: str
        The image's mime type.
    """
```

ccordoba12 commented on spyder/widgets/collectionseditor.py:
```diff
@@ -595,6 +595,12 @@ class BaseTableView(QTableView, SpyderConfigurationAccessor):
     sig_editor_creation_started = Signal()
     sig_editor_shown = Signal()
 
+    """
+    Emitted to request that a figure be shown.
+    First argument is the image, second argument is the mime type.
+    """
+    sig_show_figure_requested = Signal(bytes, str)
```
```suggestion
    sig_show_figure_requested = Signal(bytes, str)
    """
    This is emitted to request that a figure be shown.
    
    Parameters
    ----------
    image: bytes
        The image to show.
    mime_type: str
        The image's mime type.
    """
```

ccordoba12 commented on spyder/widgets/collectionseditor.py:
```diff
@@ -1418,11 +1426,21 @@ def oedit(self, key):
+            from IPython.core.pylabtools import print_figure
+            fig, ax = plt.subplots()
+            getattr(ax, funcname)(data[key])
+            png = print_figure(fig)
```
Shouldn't we use the format selected by users in Preferences here? I mean, not only `png` but also `jpeg` or `svg`?

jitseniesen replied:
I assume you are referring to the settings in Preferences > IPython Console > Graphics. I did not know about them, there are so many preferences! 

Yes, I can ensure that that setting is honoured, and also the other settings in the same box: the width, height, resolution and tightness. By the way, I don't see a way in which jpeg can be selected; am I missing something?

ccordoba12 replied:
Sorry, you're right. We only support png's and svg's.

ccordoba12 commented on spyder/widgets/collectionseditor.py:
```diff
@@ -1418,11 +1426,21 @@ def oedit(self, key):
+            getattr(ax, funcname)(data[key])
+            png = print_figure(fig)
+            collections_editor_widget = self.parent()
+            collections_editor_widget.sig_show_figure_requested.emit(
+                png, 'image/png')
```
I don't understand this: you added `sig_show_figure_requested` to `BaseTableView`, which ins inherited by this class. So, that signal is available here. 

In addition, you're connecting that signal to the one you declared in `CollectionsEditorWidget`. So, you should be able to write `self.sig_show_figure_requested.emit(png, 'image/png')` instead, right?

jitseniesen replied:
You're absolutely right, I am changing this now. I added the signal to `BaseTableView` after I wrote this code and I did not realize that this allowed me to do the simplification.

ccordoba12 commented on spyder/widgets/collectionseditor.py:
```diff
@@ -1446,6 +1464,12 @@ def set_filter(self, dictfilter=None):
 class CollectionsEditorWidget(QWidget):
     """Dictionary Editor Widget"""
 
+    """
+    Emitted to request that a figure be shown.
+    First argument is the image, second argument is the mime type.
+    """
+    sig_show_figure_requested = Signal(bytes, str)
```
```suggestion
    sig_show_figure_requested = Signal(bytes, str)
    """
    This is emitted to request that a figure be shown.
    
    Parameters
    ----------
    image: bytes
        The image to show.
    mime_type: str
        The image's mime type.
    """
```

ccordoba12 commented on spyder/widgets/collectionseditor.py:
```diff
@@ -1481,6 +1508,12 @@ def get_title(self):
 class CollectionsEditor(BaseDialog):
     """Collections Editor Dialog"""
 
+    """
+    Emitted to request that a figure be shown.
+    First argument is the image, second argument is the mime type.
+    """
+    sig_show_figure_requested = Signal(bytes, str)
```
```suggestion
    sig_show_figure_requested = Signal(bytes, str)
    """
    This is emitted to request that a figure be shown.
    
    Parameters
    ----------
    image: bytes
        The image to show.
    mime_type: str
        The image's mime type.
    """
```

ccordoba12 reviewed (commented):
Thanks @jitseniesen for your work on this!

ccordoba12 commented:
> One niggle is that plots from the collection editor will be shown in the currently active figure browser in the Plots pane, which is associated to the currently active console. However, the variable shown in the collection editor may originate from a different console.

We could set the `shellwidget` associated to the current namespace browser in the Variable Explorer as an attribute to the collection editors opened by the user from it. Then, when plotting, we could set that `shellwidget` as the focused one in the console so the plot is added to the right figure browser.

Since that would represent a potentially confusing UX for users in the future, would it be possible to address it here?

jitseniesen commented:
@ccordoba12 It is certainly possible to do as you say, and (probably) quite easy, so I have no issue with doing it in this PR.

The disadvantage is that the plot will be hidden. Suppose that the user opens one console, opens a collections editor for one of the variables in that console, then opens another console and plots from the collections editor in the first console. You say (and I agree that is most logical) that the plot should appear in the figure browser associated with the first console. However, the figure browser associated with the second console is currently shown, so the user will not get any visual feedback on their plot command.

I honestly don't know what is best.

Another aspect is that if the collections editor is coupled to the shellwidget, further down the line we can update the values in the editor if the variable in the shell changes. I think this has been requested several times; or maybe this was only for the array and dataframe editor, but the principle is similar. I would argue this is out of scope for this PR, but if this is a direction we want to move into then that's something to take into concern.

jitseniesen commented:
I made the following changes after our discussions and rebased the branch:
* Change the condition for determining whether plots should go in the Plots plugin or in a separate window. Plots will be displayed in the Plots plugin if the following three conditions are all met:
   * the Plots plugin is enabled, and
   * the setting "Mute inline plotting" in the Plots plugin is set, and
   * the graphics backend in the IPython Console preferences is set to "inline".
* When plotting in the Plots plugin, raise the plugin and also the Spyder window so that the user will see the plot immediately (I believe that a further change is needed on MS Windows for this to work because the collection editor window is not properly independent from the main Spyder window).
* When plotting in the Plots plugin, activate the console that the variable being plotted originates from. This brings the figure browser associated with that console to the front. The plot will be shown in that figure browser.
* When plotting in the Plots plugin, honour the user preferences for inline plotting as set in IPython Console > Graphics (whether to use PNG or SVG, the figure resolution, the figure size, and whether to use a tight layout).

I am not very satisfied with how I implemented the condition that the Plots plugin is enabled. As far as I know, this can only be queried in the plugin, not in a widget. I moved the plot function from the collection editor to the namespace browser, so that it is closer to the plugin level, and then wrote the following:

```python
class NamespaceBrowser:
    ...
    def plot(self, data, funcname):
        stacked_widget = self.parent()
        variable_explorer_widget = stacked_widget.parent()
        variable_explorer_plugin = variable_explorer_widget.get_plugin()
        if (variable_explorer_plugin.is_plugin_enabled(Plugins.Plots)
            ...
```

This going up to the plugin seems a bit iffy to me, but I don't also don't like any of the two alternatives that I thought of:
* I can move the whole plot function to the plugin. That is the easiest to implement but my feeling is that the plugin should only contain glue code and that the functionality should be implemented at the widget level.
* I can raise a signal, say `check_plots_plugin_enabled`. That signal propagates to the Variable Explorer plugin, which does the check and then calls yet another function at the widget level with the result of the check; that function does the actual plotting. This is perhaps the cleanest in terms of loose coupling but it requires more code and all these functions remind me of callback hell in JavaScript.

I'd be grateful for any advice on how widgets can best request information from plugins.

jitseniesen commented:
> I am not very satisfied with how I implemented the condition that the Plots plugin is enabled.

After we discussed this a few days ago, I changed the implementation (see last commit).

ccordoba12 commented on spyder/plugins/variableexplorer/widgets/namespacebrowser.py:
```diff
@@ -432,3 +448,69 @@ def save_namespace(self, filename):
+          to "inline",
+        then call `plot_in_plots_plugin`, else call `plot_in_window`.
+        """
+        if (self.plots_plugin_enabled
+                and self.get_conf('mute_inline_plotting', section='plots')
+                and self.get_conf('pylab/backend',
+                                  section='ipython_console') == 0):
```
```suggestion
        if (
            self.plots_plugin_enabled
            and self.get_conf('mute_inline_plotting', section='plots')
            and self.get_conf('pylab/backend', section='ipython_console') == 0
        ):
```

Style improvement.

ccordoba12 commented on spyder/plugins/variableexplorer/widgets/namespacebrowser.py:
```diff
@@ -432,3 +448,69 @@ def save_namespace(self, filename):
+        """
+        Plot data in Plots plugin.
+
+        Plot the given data to a PNG image and show the plot in the Plots
+        plugin.
```
```suggestion
        Plot the given data to a PNG or SVG image and show the plot in the
        Plots plugin.
```

ccordoba12 commented on spyder/plugins/variableexplorer/widgets/tests/test_namespacebrowser.py:
```diff
@@ -9,14 +9,15 @@
 from qtpy.QtCore import Qt, QPoint, QModelIndex
 
 # Local imports
+from spyder.config.manager import CONF
```
```suggestion
```

No need of this (see below).

ccordoba12 commented on spyder/plugins/variableexplorer/widgets/tests/test_namespacebrowser.py:
```diff
@@ -203,5 +204,51 @@ def test_filtering_with_large_rows(namespacebrowser, qtbot):
+    Test that plotting a list from the namespace browser sends a signal
+    with the plot if `mute_inline_plotting` is set to `True`.
+    """
+    CONF.set('plots', 'mute_inline_plotting', True)
```
```suggestion
    namespacebrowser.set_conf('mute_inline_plotting', True, section='plots')
```

`NamespaceBrowser` has `set_conf`/`get_conf`.

ccordoba12 commented on spyder/plugins/variableexplorer/widgets/tests/test_namespacebrowser.py:
```diff
@@ -203,5 +204,51 @@ def test_filtering_with_large_rows(namespacebrowser, qtbot):
+    Test that plotting a list from the namespace browser shows a plot if
+    `mute_inline_plotting` is set to `False`.
+    """
+    CONF.set('plots', 'mute_inline_plotting', False)
```
```suggestion
    namespacebrowser.set_conf('mute_inline_plotting', False, section='plots')
```

ccordoba12 commented on spyder/widgets/collectionseditor.py:
```diff
@@ -1315,10 +1315,11 @@ def paste(self):
 
-    def __init__(self, parent, data, readonly=False, title="",
-                 names=False):
+    def __init__(self, parent, data, namespacebrowser=None,  readonly=False,
```
```suggestion
    def __init__(self, parent, data, namespacebrowser=None, readonly=False,
```

ccordoba12 commented:
> After we discussed this a few days ago, I changed the implementation (see last commit).

Looks good to me now, thanks!

ccordoba12 commented on spyder/widgets/tests/test_collectioneditor.py:
```diff
@@ -922,5 +923,21 @@ def test_dicts_natural_sorting_mixed_types():
+    qtbot.addWidget(cew)
+
+    cew.editor.plot('list', 'plot')
+
```
```suggestion
```

ccordoba12 reviewed (commented):
Last suggestions, then this should be ready.

ccordoba12 reviewed (approved):
This is a really cool improvement! Thanks @jitseniesen for your work on it!!

impact27 commented:
@ccordoba12 @jitseniesen I think this PR doesn't work with py3.8. The py3.8 tests from this PR have failed (see e.g. https://github.com/spyder-ide/spyder/actions/runs/6013236935/job/16310283258)

```
E     File "/home/runner/work/spyder/spyder/spyder/plugins/variableexplorer/widgets/tests/test_namespacebrowser.py", line 222
[55](https://github.com/spyder-ide/spyder/actions/runs/6013236935/job/16310283258#step:15:56)
E       return_value=mock_png) as mock_print_figure,
[56](https://github.com/spyder-ide/spyder/actions/runs/6013236935/job/16310283258#step:15:57)
E                              ^
[57](https://github.com/spyder-ide/spyder/actions/runs/6013236935/job/16310283258#step:15:58)
E   SyntaxError: invalid syntax
```

jitseniesen commented:
@impact27 Thanks, looks like the `with (...):` syntax that I used was only introduced in Python 3.10. I'll fix this in a couple of days.
